### ISSUE TITLE
Decimal custom fields do not retain values greater than 999.99 when editing a saved transaction

### ISSUE THREAD

**Issue by tactilis** [moneymanagerex/moneymanagerex#4921]
MMEX version:
 - [x] 1.5.18

Operating System:
 - [x] Windows
 - [x] Mac OSX
 - [?] Linux 

# Description of the bug

If a Decimal custom field is given a value greater than 999.99 then, although the value is correctly saved and displayed in the transaction list, the custom field value is lost when the transaction edit dialog is reopened.

The issue was originally reported here [https://forum.moneymanagerex.org/viewtopic.php?p=23213](https://forum.moneymanagerex.org/viewtopic.php?p=23213)

# Reproduction

Is the bug reproducible? 
 - [x] Always 
 - [ ] Randomly 
 - [ ] Happened only once

Reproduction steps:

1. Define a Decimal custom field and assign it to UDFC1
Note: I set the Scale to 2 but it doesn't actually matter what that setting is.
2. Create a new deposit or withdrawal transaction with the custom field ticked and a value of 1000.00 (or greater)
3. OK the transaction.
4. Note that the custom field value of 1000.00 is correctly shown on the transaction list
5. Edit the transaction. Note that the custom field is now _not_ ticked and has a value of 0.00 **BUG**
6. OK the transaction. The custom field value is no longer shown in the transaction list

If you repeat steps 2 to 6 with a custom field value of 999.99 (or lower) the custom field value _is_ retained when you edit the transaction

Expected result:

Decimal custom field values should be retained regardless of value.

Actual result:

Decimal custom field values >= 1000 are not retained.

**Comment by whalley** [comment]
The problem here is that we use %f to format the string for the custom field value that gets stored, this will include thousands and decimal separators, when pulling the value back a ToDouble(&value) is used that fails due to the thousands separator. 

I suspect we should have just stored the value without any locale specific data. We could try to remove any locale specific data and then just treat a decimal number.

Any views on best options here?

**Comment by whalley** [comment]
Above fix feels like a bit of a bodge, but I think should work

**Comment by tactilis** [comment]
I'm afraid that won't work Mark.

Using build:
 

![image](https://user-images.githubusercontent.com/4060092/184669608-fa6f71b7-b75a-4136-9036-17b871526ddc.png)


Consider other locales. Here we have a comma used for the decimal separator:
 
![image](https://user-images.githubusercontent.com/4060092/184625208-727cd50e-efd6-4978-8583-5daa3b1c72d3.png)
 
And with a Decimal custom field:
  
![image](https://user-images.githubusercontent.com/4060092/184625401-6ff78d39-3a64-43a2-8081-8772626a6a34.png)
 
 
This is what happens:
 
https://user-images.githubusercontent.com/4060092/184669072-0b5e64e6-18a6-45f3-99e5-8e0acf0792b5.mp4

**Comment by whalley** [comment]
It's the space as thousands separator that is the issue, the decimal as . or , is handled.

This should do it: https://regex101.com/r/rCiz5g/1

**Comment by vomikan** [comment]
Better to use the following regexp: ```[,\s\.](?=\d*[,\s\.])```

> This should do it: https://regex101.com/r/rCiz5g/1

**Comment by whalley** [comment]
> Better to use the following regexp: `[,\s\.](?=\d*[,\s\.])`

Lot's of regexp variations.... But what you offers replaces the decimal point also, unless I'm doing something wrong. I wanted to keep the decimal point

**Comment by tactilis** [comment]
> It's the space as thousands separator that is the issue, the decimal as . or , is handled.
> This should do it: https://regex101.com/r/rCiz5g/1

Not quite right yet I'm afraid.

I've switched back to English(UK) in Windows Regional Settings and en_GB in MMEX options.

This incorrect formatting (no commas) occurs in the custom field when I OK the Edit Transaction dialog a **second time**, having made no changes.
 
https://user-images.githubusercontent.com/4060092/184700629-1f036310-bfb7-4e38-87cc-8dabf91c87b4.mp4

**Comment by tactilis** [comment]
Yet more odd formatting.

This time, too many decimal places after turning an **existing** custom field value off, then on again, then OK-ing the transaction.
 
https://user-images.githubusercontent.com/4060092/184705281-6a34f76d-c56c-474e-8643-eee8895db084.mp4

**Comment by whalley** [comment]
> Yet more odd formatting.

Yes, there are still a few funnies in this area.

> This time, too many decimal places after turning an **existing** custom field value off, then on again, then OK-ing the transaction.

Found the issue and have a fix.

**Comment by tactilis** [comment]
I'll hold off probing the behaviour any further until your fixes area available because anything new found may a consequence of the existing issues.

Will your fixes apply automatically to the other places that custom fields can be added, or does that first need the harmonisation of the various edit dialogs?

**Comment by whalley** [comment]
> 



> I'll hold off probing the behaviour any further until your fixes are available because anything new found may a consequence of the existing issues.

Give it another go now.

> Will your fixes apply automatically to the other places that custom fields can be added, or does that first need the harmonisation of the various edit dialogs?

The custom data handling is shared so should be picked up on other dialogs.

**Comment by vomikan** [comment]
I want to remind you about this case:
https://github.com/moneymanagerex/moneymanagerex/issues/2473

wxNumberFormatter does not know how to form a пкщгз separator for Indian localization.

**Comment by whalley** [comment]
> I want to remind you about this case: #2473
> 
> wxNumberFormatter does not know how to form a group separator for Indian localization.

Perhaps it simpler to just use below, that way we use a common set of logic in the app.
`Model_Currency::toString(value, NULL, DigitScale)`

**Comment by tactilis** [comment]
> Perhaps it simpler to just use below, that way we use a common set of logic in the app.
Model_Currency::toString(value, NULL, DigitScale)

Tested:
• Version: 1.5.19-Beta.2 64-bit
• Build on Aug 16 2022 14:59:16
• Git commit: b70b9f5 (2022-08-16)

Were you no expecting the decimal custom field to be formatted like a currency field?

Does not happen in all cases...
 
![image](https://user-images.githubusercontent.com/4060092/184924092-985c7e55-ff2a-4076-9ab3-0b4231262c47.png)

**Comment by whalley** [comment]
It will use the base currency format, this seems fine to me.

We have no idea what the user is using the field for, it may be currency it may not be. I was even tempted at one point NOT to use any currency notation and just present the raw number.

We just define the field as a DECIMAL with a specific precision. They could be recording the length in meters of string they are selling on Ebay :-)

**Comment by tactilis** [comment]
> It will use the base currency format, this seems fine to me.

Not convinced it does. Or am I not understanding something, perhaps.

![image](https://user-images.githubusercontent.com/4060092/184933058-45115bf9-8119-489f-ac75-0f9fdadc4144.png)

> We have no idea what the user is using the field for, it may be currency it may not be.
> We just define the field as a DECIMAL with a specific precision. They could be recording the length in meters of string they are selling on Ebay :-)

Yes, I agree. But surely it should respect the selected grouping char and the decimal char (either explicitly chosen or set by the locale).

In the above example, it does not.

**Comment by whalley** [comment]
> It will use the base currency format, this seems fine to me.

My mistake. It uses the defined locale, falling back to base currency if no locale.

However...... the spin control that is used seems to have its own representation that appears to always have comma as the thousands separator and colon as the decimal point. I can't see anyway to control this.

**Comment by whalley** [comment]
Just checked and the spin control converts to text based on
`wxNumberFormatter::ToString(val, m_digits);` 
using https://docs.wxwidgets.org/3.0/classwx_number_formatter.html#a42c2c2129c5bf3693e6a3699bae8595f
So we have no real control over the format used. 

Might be best to use the mmTextCtrl as used in transaction Amount fields. Probably best overall if we want to treat this as a currency type field.

**Comment by tactilis** [comment]
> the spin control that is used seems to have its own representation

An up/down spin control has little (if any) benefit in a decimal entry custom field. The arrows increment/decrement the least significant digit.  I can't see any practical use for that.

> Might be best to use the mmTextCtrl as used in transaction Amount fields. Probably best overall if we want to treat this as a currency type field.

Agreed.  Treat as a currency type (with the emphasis on _type_). i.e. formatted like currency but able to be used for any decimal purpose (like your lengths of string on ebay).

**Comment by whalley** [comment]
@tactilis Can you test this please Martin

**Comment by tactilis** [comment]
I have tried different locales. The formatting works OK now :+1: 

I have a few observations though, some of which relate to existing MMEX behaviour, which perhaps should be changed. I can break these out into separate issues if you wish.

1. The decimal field is right aligned in the edit dialog but is left aligned in the transaction list. Is it possible to right align it in the list so that the decimal points are always vertically aligned?

![image](https://user-images.githubusercontent.com/4060092/185245894-9f06e0e7-0d2c-4d53-ad38-3a2ac65009ab.png)
 
 
2. Formatting of the newly entered/edited decimal field on the Edit dialog occurs when pressing Return but does not occur when changing focus by tabbing out of the field or clicking in another field. For example:

Enter this...
 
![image](https://user-images.githubusercontent.com/4060092/185341227-7e7bd091-9e63-4ed8-9ebb-2fbe231fc76b.png)

Press Return and the field is formatted...
 
![image](https://user-images.githubusercontent.com/4060092/185340260-48ec912a-d97d-4e1f-a015-8dc86e637684.png)

But if you tab, then there is no re-formatting of the entered digits...
 
![image](https://user-images.githubusercontent.com/4060092/185340028-9a4feb59-c8b2-419d-a67a-f79d0748bfa1.png)

Ideally, the formatting of the value should occur when the field loses focus. 

3. I like that, as a consequence of your changes, the decimal field now allows simple maths. However, it no longer allows negative values (it used to). You get a validation error when OK is clicked:
 
![image](https://user-images.githubusercontent.com/4060092/185343440-b08092ad-b799-4efa-8420-ed759e7ed58b.png)

**Comment by whalley** [comment]
@tactilis Thanks for the comments...

> 1. The decimal field is right aligned in the edit dialog but is left aligned in the transaction list. Is it possible to right align it in the list so that the decimal points are always vertically aligned?

Think is is already covered here: https://github.com/moneymanagerex/moneymanagerex/issues/4868, feel free to add any extra observations to that.

> 2. Formatting of the newly entered/edited decimal field on the Edit dialog occurs when pressing Return but does not occur when changing focus by tabbing out of the field or clicking in another field. 

Will fix.

> 3. I like that, as a consequence of your changes, the decimal field now allows simple maths. However, it no longer allows negative values (it used to).

Will fix.

**Comment by vomikan** [comment]
> > 3\. Formatting of the newly entered/edited decimal field on the Edit dialog occurs when pressing Return but does not occur when changing focus by tabbing out of the field or clicking in another field.
> 
> Will fix.

@whalley I think here you can follow a path similar to mmComboBox. You can add an event directly to the mmTextCtrl class. Thus, it will be possible to simplify the code in a number of dialogs.

![image](https://user-images.githubusercontent.com/6836805/185347080-ad72e121-9259-4188-b6fb-9a6008db2398.png)

**Comment by whalley** [comment]
> Thus, it will be possible to simplify the code in a number of dialogs.

Suggest we consider any simplification in other dialogs as a separate issue, perhaps look to investigate in later release. The changes here should not detrimentally affect existing dialogs.

**Comment by vomikan** [comment]
> Suggest we consider any simplification in other dialogs as a separate issue, perhaps look to investigate in later release. The changes here should not detrimentally affect existing dialogs.

It shouldn't be difficult. This release should include.
```
wxBEGIN_EVENT_TABLE(mmTransDialog, wxDialog)
    EVT_CHAR_HOOK(mmTransDialog::OnComboKey)
    EVT_CHILD_FOCUS(mmTransDialog::OnFocusChange)

void mmTransDialog::OnFocusChange(wxChildFocusEvent& event)
...
    case mmID_TEXTAMOUNT:
        if (m_textAmount->Calculate(Model_Currency::precision(m_trx_data.ACCOUNTID))) {
            m_textAmount->GetDouble(m_trx_data.TRANSAMOUNT);
        }
        skip_amount_init_ = false;
        break;
```

Just double Calculate. This is probably a small problem.

**Comment by whalley** [comment]
It's not the fact it is difficult, it's that there are a lot of areas to review/test if we optimise. There is quite a lot of reuse of the mmTextCtrl , which is good - Account, Asset, recurring, budget, currency, filter, shares, splits, stocks, transactions. And we could tidy-up each of these... This was why I thought it best to bring out as a separate issue.

As things stand I think everything is now working (perhaps with improvements in certain dialogs where the kill focus recalculation was not done before), but there is indeed some redundancy.

**Comment by tactilis** [comment]
Tested formatting on focus changes and negative values. All good.

**Comment by whalley** [comment]
One issue I spotted is that now the mmTextCtrl control handles the text enter and focus change events internally the checkbox does not automatically get ticked when numbers are entered.... Will need to fix that.
